evm/vm/forks/sharding/vmc_handler.py

```diff
@@ -1,34 +1,145 @@
 import logging
 
+import rlp
+
 from web3.contract import (
     Contract,
 )
 
 from eth_utils import (
+    event_signature_to_log_topic,
     is_canonical_address,
     to_checksum_address,
     to_dict,
+    to_tuple,
+)
+
+from evm.rlp.sedes import (
+    address,
+    hash32,
 )
 
 from evm.utils.hexadecimal import (
+    encode_hex,
     decode_hex,
 )
+from evm.utils.numeric import (
+    big_endian_to_int,
+)
 
 from evm.vm.forks.sharding.config import (
     get_sharding_config,
 )
 
 
+class NextLogUnavailable(Exception):
+    pass
+
+
+@to_dict
+def parse_collation_added_data(data_hex):
+    data_bytes = decode_hex(data_hex)
+    score = big_endian_to_int(data_bytes[-32:])
+    is_new_head = bool(big_endian_to_int(data_bytes[-64:-32]))
+    header_bytes = data_bytes[:-64]
+    # [num, num, bytes32, bytes32, bytes32, address, bytes32, bytes32, num, bytes]
+    sedes = rlp.sedes.List([
+        rlp.sedes.big_endian_int,
+        rlp.sedes.big_endian_int,
+        hash32,
+        hash32,
+        hash32,
+        address,
+        hash32,
+        hash32,
+        rlp.sedes.big_endian_int,
+        rlp.sedes.binary,
+    ])
+    header_values = rlp.decode(header_bytes, sedes=sedes)
+    yield 'header', header_values
+    yield 'is_new_head', is_new_head
+    yield 'score', score
+
+
 class VMC(Contract):
 
-    logger = logging.getLogger("evm.chain.sharding.mainchain_handler.VMC")
+    logger = logging.getLogger("evm.chain.sharding.VMC")
+
+    # For handling logs filtering
+    # Event:
+    #   CollationAdded(indexed uint256 shard, bytes collationHeader, bool isNewHead, uint256 score)
+    COLLATION_ADDED_TOPIC = event_signature_to_log_topic(
+        "CollationAdded(int128,bytes4096,bool,int128)"
+    )
 
-    def __init__(self, *args, default_privkey, **kwargs):
+    shards = set()
+    # shard_id -> list
+    # older <---------------> newer
+    new_collation_added_logs = {}
+    # shard_id -> list
+    # newer <---------------> older
+    unchecked_collation_added_logs = {}
+    # shard_id -> score
+    current_checking_score = {}
+
+    def __init__(self, *args, log_handler, default_privkey, **kwargs):
+        self.log_handler = log_handler
         self.default_privkey = default_privkey
         self.default_sender_address = default_privkey.public_key.to_canonical_address()
         self.config = get_sharding_config()
         super().__init__(*args, **kwargs)
 
+    def init_shard_variables(self, shard_id):
+        self.shards.add(shard_id)
+        self.unchecked_collation_added_logs[shard_id] = []
+        self.new_collation_added_logs[shard_id] = []
+        self.current_checking_score[shard_id] = None
```

Comment:
I think that you can remove all of this initialization if you use `collection.defaultdict` instances for all of these three data structures.

```python
self.unchecked_collation_added_logs = collections.defaultdict(list)
self.new_collation_added_logs = collections.defaultdict(list)
self.current_checking_score = {}
```

Then you can access both `self.unchecked_collation_added_logs[shard_id]` and `self.new_collation_added_logs[shard_id]` as if they have already been initialized thanks to the magic of defaultdicts.

This should allow you to remove all of the initialization and `ensure_` initialization code.

Reply:
It's a good idea. Thank you for pointing out. I will change it.